spec/index.bs

```diff
@@ -2517,6 +2518,82 @@ dictionary GPUImageBitmapCopyView {
+
+<div algorithm class=validusage>
+
+<dfn abstract-op>copyBufferToTexture Valid Usage</dfn>
+
+Given a {{GPUCommandEncoder}} |encoder| and the arguments {{GPUBufferCopyView}} |source|, {{GPUTextureCopyView}} |destination|, {{GPUExtent3D}} |copySize|, the following validation rules apply:
+
+  - |encoder| must be a [=valid=] {{GPUCommandEncoder}}.
+  - |encoder|.{{GPUCommandEncoder/copyBufferToTexture()}} must not be called when a {{GPURenderPassEncoder}} is active on |encoder|.
+  - |encoder|.{{GPUCommandEncoder/copyBufferToTexture()}} must not be called when a {{GPUComputePassEncoder}} is active on |encoder|.
+  - |source|.{{GPUBufferCopyView/buffer}} must be a [=valid=] {{GPUBuffer}}.
+  - |destination|.{{GPUTextureCopyView/texture}} must be a [=valid=] {{GPUTexture}}.
+  - The {{GPUBuffer/[[usage]]}} of |source|.{{GPUBufferCopyView/buffer}} must contain {{GPUBufferUsage/COPY_SRC}}.
+  - The {{GPUTexture/[[textureUsage]]}} of |destination|.{{GPUTextureCopyView/texture}} must contain {{GPUTextureUsage/COPY_DST}}.
+  - The {{GPUTexture/[[sampleCount]]}} of |destination|.{{GPUTextureCopyView/texture}} must be 1.
+  - If |source|.{{GPUBufferCopyView/rowsPerImage}} is not 0, it must be greater than or equal to |copySize|.{{GPUExtent3DDict/height}}.
+  - |source|.{{GPUBufferCopyView/rowsPerImage}} must be a multiple of [=texel block height=] of the {{GPUTexture/[[format]]}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - |destination|.{{GPUTextureCopyView/origin}}.{{GPUOrigin3DDict/x}} must be a multiple of [=texel block width=] of the {{GPUTexture/[[format]]}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - |destination|.{{GPUTextureCopyView/origin}}.{{GPUOrigin3DDict/y}} must be a multiple of [=texel block height=] of the {{GPUTexture/[[format]]}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - |copySize|.{{GPUExtent3DDict/width}} must be a multiple of [=texel block width=] of the {{GPUTexture/[[format]]}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - |copySize|.{{GPUExtent3DDict/height}} must be a multiple of [=texel block height=] of the {{GPUTexture/[[format]]}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - |source|.{{GPUBufferCopyView/bytesPerRow}} must be a multiple of 256.
+  - If |source|.{{GPUBufferCopyView/bytesPerRow}} is not 0, it must be greater than or equal to the number of |bytes in a complete row|.
+  - If |source|.{{GPUBufferCopyView/bytesPerRow}} is 0, the number of |bytes in a complete row| must be a multiple of 256.
+  - |destination|.{{GPUTextureCopyView/mipLevel}} must be less than the {{GPUTexture/[[mipLevelCount]]}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - |destination|.{{GPUTextureCopyView/arrayLayer}} must be less than the {{GPUTexture/[[textureSize]]}}.{{GPUExtent3DDict/depth}} of |destination|.{{GPUTextureCopyView/texture}}.
+  - {{GPUExtent3DDict/depth}} must be 1 when the {{GPUTexture/[[dimension]]}} of {{GPUTextureCopyView/texture}} is {{GPUTextureDimension/2d}}.
```

Comment:
also, for `GPUTextureDimension/1d`

Reply:
Actually not true for 2d textures, because all 2d textures are semantically 2d texture arrays. Instead, this should just say the depth must be <= the texture's depth.

Reply:
Metal only supports copying into one layer of a 2d texture:
https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400752-copyfrombuffer?language=objc
"when you copy to a 2D texture, depth must be 1."

Reply:
I think we want to expose multi-layer copies, we can implement them as multiple copies when necessary

Reply:
> "when you copy to a 2D texture, depth must be 1."

I'm pretty sure this is talking about 2D non-array textures specifically.

That documentation is kind of confusing though. On sourceBytesPerImage it says:

> For 3D textures and 2D array textures, the stride in bytes between 2D images of the source buffer memory. The value must be a multiple of the destination texture's pixel size, in bytes.

which strongly implies that copies into multiple layers of a 2D array texture is allowed. But destinationSlice doesn't say anything about it. I think destinationSlice is actually destinationFirstSlice?

Reply:
Hi @JusSn @litherum, what do you think?

I also notice that in MoltenVK the multi-layer copy is implemented by [an iteration over every layers in a for-loop](https://github.com/KhronosGroup/MoltenVK/blob/master/MoltenVK/MoltenVK/Commands/MVKCmdTransfer.mm#L864).